### PR TITLE
First attempt at autocompletion for CrossRef

### PR DESCRIPTION
**Pull request by Siedlerchr**
Fixes https://github.com/koppor/jabref/issues/257
Works partly:

grafik

TODO: Make it appear as Tag-Item and not as textinput. I currently have no idea 
I hope @tobiasdiez  has an idea how to do it

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [ ] Screenshots added (for bigger UI changes)
- [x] Manually tested changed features in running JabRef
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)
- [ ] If you changed the localization: Did you run `gradle localizationUpdate`?

### CODE REVIEW

**Comment by koppor** [comment]
I got a minor issue: In the default setting of JabRef, no auto completion happens.

grafik

Option A: Add the list of fields where we support referencing other entries to that list.
Option B: Always do autocompletion in that case

I vote for B, because that automcompletion "only" references BibTeX keys and we have no issues with name rewriting etc.

**Comment by Siedlerchr** [comment]
@lenhard  You have to press 2 times enter: First enter the autcomplete value is taken in the field, second time enter the tag field rendering is created from it and then it's  also visible in the bibtex source tab and therefore in the database. So

**Comment by lenhard** [comment]
@Siedlerchr A thanks, it works as you said. 

I guess it would be cool if JabRef somehow lets you know that the crossref data is not really complete and will be deleted. Just to avoid this surprise. But I don't know how this has been before and hence the current solution is good enough for me as well.

**Comment by Siedlerchr** [comment]
@lenhard  Yes, that's a bit weird, but I have no idea how to solve this. I will add a message in the changelog

**Comment by Siedlerchr** [comment]
The only thing which is currently not implemented is Jump to entry which seems to be not possible when I look at the code comment. Otherwise it's ready to go from my side

File: src/main/java/org/jabref/gui/autocompleter/AutoCompletePreferences.java [hunk 1]
```diff
@@ -13,15 +13,15 @@
...
+    private List<String> completeFields;
+    private final JournalAbbreviationPreferences journalAbbreviationPreferences;
 
     public AutoCompletePreferences(boolean shouldAutoComplete, AutoCompleteFirstNameMode firstNameMode, boolean onlyCompleteLastFirst, boolean onlyCompleteFirstLast, List<String> completeNames, JournalAbbreviationPreferences journalAbbreviationPreferences) {
         this.shouldAutoComplete = shouldAutoComplete;
         this.firstNameMode = firstNameMode;
         this.onlyCompleteLastFirst = onlyCompleteLastFirst;
         this.onlyCompleteFirstLast = onlyCompleteFirstLast;
-        this.completeNames = completeNames;
+        this.completeFields = completeNames;
```
Comment (koppor) [inline src/main/java/org/jabref/gui/autocompleter/AutoCompletePreferences.java]: Why not renaming the parameter, too? Maybe `fieldsToComplete` would be a good name for both variables? `completeFields` is also OK for me.

File: src/main/java/org/jabref/gui/autocompleter/ParsedEntryLinkSuggestionProvider.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    protected Comparator<ParsedEntryLink> getComparator() {
+        return (o1, o2) -> o1.getKey().compareTo(o2.getKey());
+    }
+
+    @Override
+    protected boolean isMatch(ParsedEntryLink suggestion, ISuggestionRequest request) {
+        String userTextLower = request.getUserText().toLowerCase(Locale.ROOT);
+        return suggestion.getKey().toLowerCase(Locale.ROOT).contains(userTextLower);
+    }
+}
```
Comment (koppor) [inline src/main/java/org/jabref/gui/autocompleter/ParsedEntryLinkSuggestionProvider.java]: Could you quickly add a EOF newline?

https://marketplace.eclipse.org/content/editorconfig-eclipse :)

File: src/main/java/org/jabref/gui/autocompleter/SuggestionProviders.java [hunk 3]
```diff
@@ -60,7 +60,7 @@ public void indexEntry(BibEntry bibEntry) {
         if (InternalBibtexFields.getFieldProperties(fieldName).contains(FieldProperty.PERSON_NAMES)) {
             return new PersonNameSuggestionProvider(fieldName);
         } else if (InternalBibtexFields.getFieldProperties(fieldName).contains(FieldProperty.SINGLE_ENTRY_LINK)) {
-            return new BibEntrySuggestionProvider();
+            return new ParsedEntryLinkSuggestionProvider();
```
Comment (koppor) [inline src/main/java/org/jabref/gui/autocompleter/SuggestionProviders.java]: Could you comment, why the `BibEntrySuggestionProvider()` cannot be used here? Maybe just add a comment before this line.

File: src/main/java/org/jabref/gui/autocompleter/AutoCompletePreferences.java [hunk 4]
```diff
@@ -13,15 +13,15 @@
     private AutoCompleteFirstNameMode firstNameMode;
     private boolean onlyCompleteLastFirst;
     private boolean onlyCompleteFirstLast;
-    private List<String> completeNames;
-    private JournalAbbreviationPreferences journalAbbreviationPreferences;
+    private List<String> completeFields;
+    private final JournalAbbreviationPreferences journalAbbreviationPreferences;
 
     public AutoCompletePreferences(boolean shouldAutoComplete, AutoCompleteFirstNameMode firstNameMode, boolean onlyCompleteLastFirst, boolean onlyCompleteFirstLast, List<String> completeNames, JournalAbbreviationPreferences journalAbbreviationPreferences) {
```
Comment (lenhard) [inline src/main/java/org/jabref/gui/autocompleter/AutoCompletePreferences.java]: Nitpick: If you rename this to `completeFields` then can you please also rename the input parameter if the constructor?

File: src/main/java/org/jabref/gui/autocompleter/AutoCompletePreferences.java [hunk 5]
```diff
@@ -60,20 +60,24 @@ public void setOnlyCompleteFirstLast(boolean onlyCompleteFirstLast) {
...
+        return completeFields;
     }
 
-    public void setCompleteNames(List<String> completeNames) {
-        this.completeNames = completeNames;
+    public void setCompleteFields(List<String> completeFields) {
+        this.completeFields = completeFields;
     }
 
     public void setCompleteNames(String input) {
```
Comment (lenhard) [inline src/main/java/org/jabref/gui/autocompleter/AutoCompletePreferences.java]: Same as above: Please fix the method name: names -> fields

File: src/main/java/org/jabref/gui/autocompleter/AutoCompletePreferences.java [hunk 6]
```diff
@@ -60,20 +60,24 @@ public void setOnlyCompleteFirstLast(boolean onlyCompleteFirstLast) {
...
-        setCompleteNames(Arrays.asList(input.split(DELIMITER)));
+        setCompleteFields(Arrays.asList(input.split(DELIMITER)));
     }
 
     public String getCompleteNamesAsString() {
```
Comment (lenhard) [inline src/main/java/org/jabref/gui/autocompleter/AutoCompletePreferences.java]: Same as above: Please fix the method name: names -> fields

File: src/main/java/org/jabref/preferences/JabRefPreferences.java [hunk 7]
```diff
@@ -581,7 +584,7 @@ private JabRefPreferences() {
         defaults.put(AUTOCOMPLETER_FIRSTNAME_MODE, AutoCompleteFirstNameMode.BOTH.name());
         defaults.put(AUTOCOMPLETER_FIRST_LAST, Boolean.FALSE); // "Autocomplete names in 'Firstname Lastname' format only"
         defaults.put(AUTOCOMPLETER_LAST_FIRST, Boolean.FALSE); // "Autocomplete names in 'Lastname, Firstname' format only"
-        defaults.put(AUTOCOMPLETER_COMPLETE_FIELDS, "author;editor;title;journal;publisher;keywords");
+        defaults.put(AUTOCOMPLETER_COMPLETE_FIELDS, "author;editor;title;journal;publisher;keywords;crossref");
```
Comment (koppor) [inline src/main/java/org/jabref/preferences/JabRefPreferences.java]: I think, `;related;entryset` is missing here. Can you add them please? `entryset` is from biblatex.

If someone else wants to check: `FieldProperty.SINGLE_ENTRY_LINK` and `FieldProperty.MULTIPLE_ENTRY_LINK` could be possible starting points.

**Comment by Siedlerchr** [comment]
I thought I found all fields while replacing..



Am 07.09.2017 3:47 nachm. schrieb "Oliver Kopp" <notifications@github.com>:

*@koppor* requested changes on this pull request.

some last words minor things.
------------------------------

In src/main/java/org/jabref/gui/autocompleter/AutoCompletePreferences.java
<https://github.com/JabRef/jabref/pull/3201#discussion_r137533294>:

>
     public AutoCompletePreferences(boolean shouldAutoComplete,
AutoCompleteFirstNameMode firstNameMode, boolean
onlyCompleteLastFirst, boolean onlyCompleteFirstLast, List<String>
completeNames, JournalAbbreviationPreferences
journalAbbreviationPreferences
) {
         this.shouldAutoComplete = shouldAutoComplete;
         this.firstNameMode = firstNameMode;
         this.onlyCompleteLastFirst = onlyCompleteLastFirst;
         this.onlyCompleteFirstLast = onlyCompleteFirstLast;
-        this.completeNames = completeNames;
+        this.completeFields = completeNames;

Why not renaming the parameter, too? Maybe fieldsToComplete would be a good
name for both variables? completeFields is also OK for me.
------------------------------

In src/main/java/org/jabref/gui/autocompleter/ParsedEntryLinkSuggestionProvi
der.java <https://github.com/JabRef/jabref/pull/3201#discussion_r137533510>:

> +        }
+
+        addPossibleSuggestions(new ParsedEntryLink(entry));
+    }
+
+    @Override
+    protected Comparator<ParsedEntryLink> getComparator() {
+        return (o1, o2) -> o1.getKey().compareTo(o2.getKey());
+    }
+
+    @Override
+    protected boolean isMatch(ParsedEntryLink suggestion,
ISuggestionRequest request) {
+        String userTextLower = request.getUserText().toLowerCase(Locale.ROOT);
+        return
suggestion.getKey().toLowerCase(Locale.ROOT).contains(userTextLower);
+    }
+}

Could you quickly add a EOF newline?

https://marketplace.eclipse.org/content/editorconfig-eclipse :)
------------------------------

In src/main/java/org/jabref/gui/autocompleter/SuggestionProviders.java
<https://github.com/JabRef/jabref/pull/3201#discussion_r137533708>:

> @@ -60,7 +60,7 @@ public void indexEntry(BibEntry bibEntry) {
         if (InternalBibtexFields.getFieldProperties(fieldName).contains(FieldProperty.PERSON_NAMES))
{
             return new PersonNameSuggestionProvider(fieldName);
         } else if
(InternalBibtexFields.getFieldProperties(fieldName).contains(FieldProperty.SINGLE_ENTRY_LINK))
{
-            return new BibEntrySuggestionProvider();
+            return new ParsedEntryLinkSuggestionProvider();

Could you comment, why the BibEntrySuggestionProvider() cannot be used
here? Maybe just add a comment before this line.
------------------------------

In src/main/java/org/jabref/preferences/JabRefPreferences.java
<https://github.com/JabRef/jabref/pull/3201#discussion_r137542644>:

> @@ -581,7 +584,7 @@ private JabRefPreferences() {
         defaults.put(AUTOCOMPLETER_FIRSTNAME_MODE,
AutoCompleteFirstNameMode.BOTH.name());
         defaults.put(AUTOCOMPLETER_FIRST_LAST, Boolean.FALSE); //
"Autocomplete names in 'Firstname Lastname' format only"
         defaults.put(AUTOCOMPLETER_LAST_FIRST, Boolean.FALSE); //
"Autocomplete names in 'Lastname, Firstname' format only"
-        defaults.put(AUTOCOMPLETER_COMPLETE_FIELDS,
"author;editor;title;journal;publisher;keywords");
+        defaults.put(AUTOCOMPLETER_COMPLETE_FIELDS,
"author;editor;title;journal;publisher;keywords;crossref");

I think, ;related;entryset is missing here. Can you add them please?
entryset is from biblatex.

If someone else wants to check: FieldProperty.SINGLE_ENTRY_LINK and
FieldProperty.MULTIPLE_ENTRY_LINK could be possible starting points.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/JabRef/jabref/pull/3201#pullrequestreview-61213391>,
or mute
the thread

**Comment by Siedlerchr** [comment]
I now added the new keys to default and adjusted the migration. I also added a comment. @koppor

**Pull request merged**